src/Hydrodynamics/Hydrodynamics.cpp

```diff
@@ -123,6 +123,65 @@ Hydrodynamics<ndim>::Hydrodynamics(int hydro_forces_aux, int self_gravity_aux, F
 
 
 //=================================================================================================
+//  Hydrodynamics::CreateNewParticle
+/// Create a new hydro particle in the main arrays at the end of the existing particles.
+/// Also triggers the re-creation of ghosts and another tree-build to include new particles.
+//=================================================================================================
+template <int ndim>
+Particle<ndim>& Hydrodynamics<ndim>::CreateNewParticle
+ (const enum ptype _ptype,             ///< [in] ptype of new particle
+  const enum parttype _parttype,       ///< [in] parttype of new particle
+  const int n,                         ///< [in] Current integer time
+  const int level_step,                ///< [in] ..
+  const int level_max,                 ///< [in] Current maximum block timestep level
+  const FLOAT t,                       ///< [in] Current simulation time
+  const FLOAT m,                       ///< [in] Mass of new particle
+  const FLOAT u,                       ///< [in] Specific internal energy of new particle
+  const FLOAT r[ndim],                 ///< [in] Position of new particle
+  const FLOAT v[ndim])                 ///< [in] Velocity of new particle
+{
+  // First, check if there is space for the new particle.  Otherwisr, throw an exception.
+  if (Nhydro >= Nhydromax) {
+    ExceptionHandler::getIstance().raise("Run out of memory for new hydro particles");
```

Comment:
Sure, but I assume you mean we can reallocate and increase the arrays rather than unnecessarily killing the simulation?